onedal/neighbors/neighbors.py

```diff
@@ -0,0 +1,397 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#===============================================================================
+
+from abc import ABCMeta, abstractmethod
+from enum import Enum
+import sys
+from numbers import Number, Integral
+import warnings
+
+from distutils.version import LooseVersion
+
+import numpy as np
+from scipy import sparse as sp
+from ..datatypes import (
+    _check_X_y,
+    _check_array,
+    _check_is_fitted,
+    _column_or_1d,
+    _check_n_features,
+    _check_classification_targets,
+    _is_classifier,
+    _is_regressor
+)
+
+from onedal import _backend
+
+from ..common._mixin import ClassifierMixin, RegressorMixin
+from ..common._policy import _get_policy
+
+
+def _parse_auto_method(estimator, method, n_samples, n_features):
+    result_method = method
+
+    if (method in ['auto', 'ball_tree']):
+        condition = estimator.n_neighbors is not None and \
+            estimator.n_neighbors >= estimator.n_samples_fit_ // 2
+        if estimator.metric == 'precomputed' or n_features > 11 or condition:
+            result_method = 'brute'
+        else:
+            if estimator.metric in 'euclidean':
+                result_method = 'kd_tree'
+            else:
+                result_method = 'brute'
+
+    return result_method
+
+def _validate_data(estimator, X, y=None, reset=True,
+                  validate_separately=False, **check_params):
+    if y is None:
+        try:
+            requires_y = estimator._get_tags()["requires_y"]
+        except KeyError:
+            requires_y = False
+
+        if requires_y:
+            raise ValueError(
+                f"This {estimator.__class__.__name__} estimator "
+                f"requires y to be passed, but the target y is None."
+            )
+        X = _check_array(X, **check_params)
+        out = X, y
+    else:
+        if validate_separately:
+            # We need this because some estimators validate X and y
+            # separately, and in general, separately calling _check_array()
+            # on X and y isn't equivalent to just calling _check_X_y()
+            # :(
+            check_X_params, check_y_params = validate_separately
+            X = _check_array(X, **check_X_params)
+            y = _check_array(y, **check_y_params)
+        else:
+            X, y = _check_X_y(X, y, **check_params)
+        out = X, y
+
+    if check_params.get('ensure_2d', True):
+        _check_n_features(X, reset=reset)
+
+    return out
+
+def _onedal_fit(self, X, y, queue):
+    policy = _get_policy(queue, X, y)
+    params = self._get_onedal_params(X)
+    train_alg = _backend.neighbors.classification.train(policy, params,

```

Comment:
In `cpp` you create bindings for two tasks - classification and search, but here you are using just one. Why?